File: python/astroengine/turbo.py
"""astroengine.turbo -- a fast-evaluation tier fit to the engine's own output.

The full engine evaluates VSOP/ELP series (hundreds of terms) per body. The
turbo tier replaces that with a segmented Chebyshev representation of the
engine's apparent ecliptic longitude, so a longitude costs a couple of dozen
multiply-adds. A century-scale transit scan that calls longitude() tens of
thousands of times then runs in milliseconds, in the browser.

The pack is fit to the *engine* (not an external source), so the only claim is
that turbo reproduces the engine to the fit tolerance; the engine's own
accuracy vs Swiss Ephemeris / JPL is unchanged. Each segment fits the
longitude unwrapped relative to the segment centre, so a segment that straddles
0/360 has no jump.

Segment lengths must resolve the ~13.7-day semi-monthly nutation term (shared
by every body), so even the slow planets use ~24-day segments. At degree 12
that reproduces the planets to <0.01" and the Moon to ~0.03".

The fit is pure Python (Chebyshev-Gauss-Lobatto interpolation, no numpy), so
the pack is reproducible anywhere the engine runs.
"""
import json
import math
import os

from .chebyshev import _clenshaw

DEG = math.pi / 180.0


def cheb_coeffs_cgl(samples):
    """Chebyshev coefficients (degree N) interpolating `samples`, which are a
    function sampled at the N+1 Chebyshev-Gauss-Lobatto nodes x_k = cos(k*pi/N),
    k = 0..N. Pure-Python discrete cosine transform."""
    n = len(samples) - 1
    coeffs = []
    for j in range(n + 1):
        s = 0.5 * samples[0] + 0.5 * samples[n] * math.cos(math.pi * j)
        for k in range(1, n):
            s += samples[k] * math.cos(math.pi * j * k / n)
        coeffs.append((2.0 / n) * s)
    coeffs[0] *= 0.5
    coeffs[n] *= 0.5
    return coeffs


def _wrap180(d):
    return ((d + 180.0) % 360.0) - 180.0


def _fit_segment(engine, body, a, seg, degree, zodiac):
    """Chebyshev coefficients for `body`'s longitude over [a, a+seg]. The
    samples are unwrapped relative to the segment centre, so a segment that
    straddles 0/360 has no jump; the body only has to stay within 180 deg of
    the centre value across the segment (true for every body at a sane segment
    length)."""
    ref = engine.longitude(body, a + seg / 2.0, zodiac=zodiac)
    samples = []
    for k in range(degree + 1):
        x = math.cos(math.pi * k / degree)          # CGL node in [-1, 1]
        t = a + (x + 1.0) * seg / 2.0
        lon = engine.longitude(body, t, zodiac=zodiac)
        samples.append(ref + _wrap180(lon - ref))
    return cheb_coeffs_cgl(samples)


def fit(engine, bodies, jd0, jd1, seg_days, degree=12, zodiac="tropical"):
    """Fit a turbo pack: per body, segmented Chebyshev of (unwrapped) longitude.
    `seg_days` is a dict body -> segment length (days)."""
    pack = {"jd0": jd0, "jd1": jd1, "degree": degree, "zodiac": zodiac,
            "bodies": {}}
    for body in bodies:
        seg = seg_days[body]
        nseg = int(math.ceil((jd1 - jd0) / seg))
        segs = [_fit_segment(engine, body, jd0 + i * seg, seg, degree, zodiac)
                for i in range(nseg)]
        pack["bodies"][body] = {"seg_days": seg, "segments": segs}
    return pack


class Turbo:
    """Runtime evaluator for a turbo pack."""

    def __init__(self, pack):
        self.jd0 = pack["jd0"]
        self.jd1 = pack["jd1"]
        self.bodies = pack["bodies"]

    @classmethod
    def load(cls, path):
        with open(path) as f:
            return cls(json.load(f))

    def has(self, body):
        return body in self.bodies

    def longitude(self, body, jd):
        """Apparent ecliptic longitude (degrees) from the turbo pack."""
        b = self.bodies[body]
        seg = b["seg_days"]
        if not (self.jd0 <= jd <= self.jd1):
            raise ValueError(f"jd {jd} outside turbo range {self.jd0}-{self.jd1}")
        i = min(int((jd - self.jd0) / seg), len(b["segments"]) - 1)
        x = 2.0 * (jd - (self.jd0 + i * seg)) / seg - 1.0
        return _clenshaw(b["segments"][i], x) % 360.0

File: packages/caelus/src/turbo.ts
/**
 * astroengine turbo -- fast longitude evaluator for a turbo pack.
 *
 * A turbo pack is a segmented Chebyshev representation of the engine's apparent
 * longitude, fit to the engine itself (see python/astroengine/turbo.py and
 * fit_turbo.py). Evaluating a longitude is a couple of dozen multiply-adds, so
 * a century-scale transit scan that calls it tens of thousands of times runs in
 * milliseconds. The pack is data you mint for your range and bodies; this is
 * the runtime-only evaluator (no fitting, no engine, no I/O).
 *
 * The evaluator mirrors the Python reference exactly, so both reproduce the
 * pack bit-identically.
 */
import { mod } from "./core.js";

export interface TurboBody {
  seg_days: number;
  segments: number[][];
}

export interface TurboPack {
  jd0: number;
  jd1: number;
  degree: number;
  zodiac: string;
  bodies: Record<string, TurboBody>;
}

function clenshaw(coeffs: number[], x: number): number {
  let b0 = 0;
  let b1 = 0;
  for (let i = coeffs.length - 1; i >= 1; i--) {
    const t = 2 * x * b0 - b1 + coeffs[i];
    b1 = b0;
    b0 = t;
  }
  return x * b0 - b1 + coeffs[0];
}

/**
 * Runtime evaluator for a **turbo pack** — a segmented Chebyshev fit of the
 * engine's apparent longitude over a fixed range and body set. Evaluating a
 * longitude costs a couple dozen multiply-adds, so a century-scale transit scan
 * that calls it tens of thousands of times runs in milliseconds. Construct one
 * from a pack you minted offline; it does no fitting, no I/O, and needs no
 * {@link Engine}.
 *
 * @example
 * ```ts
 * const turbo = new Turbo(pack); // a TurboPack generated for your range/bodies
 * if (turbo.has("mars")) turbo.longitude("mars", jd);
 * ```
 */
export class Turbo {
  /** Start of the pack's valid Julian Day (UT) range. */
  readonly jd0: number;
  /** End of the pack's valid Julian Day (UT) range. */
  readonly jd1: number;
  private readonly bodies: Record<string, TurboBody>;

  /**
   * @param pack A {@link TurboPack}: the fitted segments plus its `jd0`/`jd1`
   *   range, minted offline for your bodies and span.
   */
  constructor(pack: TurboPack) {
    this.jd0 = pack.jd0;
    this.jd1 = pack.jd1;
    this.bodies = pack.bodies;
  }

  /**
   * Whether this pack can evaluate a given body.
   *
   * @param body Body id to test.
   * @returns `true` if {@link Turbo.longitude} accepts `body`.
   */
  has(body: string): boolean {
    return body in this.bodies;
  }

  /**
   * Apparent ecliptic longitude (degrees) of a body from the turbo pack, in the
   * pack's own zodiac. The hot path for bulk scans.
   *
   * @param body A body id the pack contains (see {@link Turbo.has}).
   * @param jd Julian Day (UT), within `[jd0, jd1]`.
   * @returns Ecliptic longitude in degrees, `[0, 360)`.
   * @throws Error if the pack lacks `body`, or `jd` is outside `[jd0, jd1]`.
   */
  longitude(body: string, jd: number): number {
    const b = this.bodies[body];
    if (!b) throw new Error(`turbo: no pack for ${body}`);
    if (jd < this.jd0 || jd > this.jd1) {
      throw new Error(`jd ${jd} outside turbo range ${this.jd0}-${this.jd1}`);
    }
    const seg = b.seg_days;
    const i = Math.min(Math.floor((jd - this.jd0) / seg), b.segments.length - 1);
    const x = 2 * (jd - (this.jd0 + i * seg)) / seg - 1;
    return mod(clenshaw(b.segments[i], x), 360);
  }
}
